libcxx/include/semaphore

```diff
@@ -94,20 +95,25 @@ public:
     auto __old = __a_.fetch_add(__update, memory_order_release);
     _LIBCPP_ASSERT_ARGUMENT_WITHIN_DOMAIN(
         __update <= _LIBCPP_SEMAPHORE_MAX - __old, "update is greater than the expected value");
-
-    if (__old > 0) {
-      // Nothing to do
-    } else if (__update > 1)
+    (void)__old;
+    if (__update > 1) {
       __a_.notify_all();
-    else
+    } else {
       __a_.notify_one();
+    }
   }
   _LIBCPP_AVAILABILITY_SYNC _LIBCPP_HIDE_FROM_ABI void acquire() {
     auto const __test_fn = [this]() -> bool {
       auto __old = __a_.load(memory_order_relaxed);
       return (__old != 0) && __a_.compare_exchange_strong(__old, __old - 1, memory_order_acquire, memory_order_relaxed);
     };
-    __cxx_atomic_wait(&__a_.__a_, __test_fn);
+    auto const __test_with_old = [this](__cxx_contention_t& __monitor) -> bool {
```

Comment:
> it would immediately become apparent that `latch` suffers from the same race

I don't think `latch` has the same problem as the none of the waiters are going to change anything. In fact, I don't understand why `latch` needs to use this complicated internal functions at all. can't it just do
```
__a_.wait(0, memory_order_acquire);
```

Reply:
> I don't think `latch` has the same problem as the none of the waiters are going to change anything. In fact, I don't understand why `latch` needs to use this complicated internal functions at all. can't it just do
> 
> ```
> __a_.wait(0, memory_order_acquire);
> ```

I think it's written in terms of `try_wait()` and `__cxx_atomic_wait` to express the separation of concerns -- the `latch` only has to implement the non-blocking logic in `try_wait()`, and `__cxx_atomic_wait` "generically" turns the `try_wait()` into a blocking algorithm. To do this, `__cxx_atomic_wait` implements the "event count" pattern:

- https://github.com/facebook/folly/blob/dd5e8b4fcef37193cd9676e1122d1621a1d90c49/folly/experimental/EventCount.h
- https://cbloomrants.blogspot.com/2011/07/07-08-11-who-ordered-event-count.html

What complicates the internal `__cxx_atomic_wait` eventcount logic a lot is that sometimes the eventcount futex is from the "parking lot" (i.e. `__libcpp_contention_table`) and other times it is the atomic variable itself if it's directly usable as a platform futex (leading to those ABA style issues).

Reply:
> * https://github.com/facebook/folly/blob/dd5e8b4fcef37193cd9676e1122d1621a1d90c49/folly/experimental/EventCount.h

but I still think 
```cpp
__a_.wait(0, memory_order_acquire);
```
is simpler and easier to understand.

w.r.t to the `latch`, do you agree that in the current form, there is no issues like the one in `semaphore`? All the modification of a `latch` is going one direction so there should not be ABA problem IIUC

Reply:
> w.r.t to the `latch`, do you agree that in the current form, there is no issues like the one in `semaphore`? All the modification of a `latch` is going one direction so there should not be ABA problem IIUC

Ah yes, you are right! The latch is a downward counter, so the `__test_fn` could only "err" in the direction of not-blocking. So if the monitor read "1" for example, the `__test_fn`  could only read either "1" as well (and then the algorithm would proceed to block), or read "0" and exit the loop.

I still think it's worth to refactor it so that the `__test_fn` always takes the current value of the atomic as an in-out-argument, as this would just make this class of ABA issue go away completely. More synchronization primitives might be implemented in terms of `__cxx_atomic_wait` in the future.

Reply:
> but I still think
> 
> ```c++
> __a_.wait(0, memory_order_acquire);
> ```
> 
> is simpler and easier to understand.

I absolutely agree -- in this case it is easier to understand on a first glance. On the other hand, in general, you don't want to even spend one nanosecond thinking about your `wait()` method. You just want to take your `try_wait()` and your eventcount abstraction (i.e. `__cxx_atomic_wait()`), compose them, and get your `wait()` for free.

So even if that code would be longer, it is simpler from an architectural viewpoint. Ideally, reviewers would see `wait()` and go "ha, this is just `try_wait()` made blocking by an eventcount". I do think though that the current code in `semaphore`/`latch` could be made clearer to express that design intent, and a few comments wouldn't have hurt, either...

Reply:
> You just want to take your `try_wait()` and your eventcount abstraction (i.e. `__cxx_atomic_wait()`), compose them, and get your `wait()` for free.

For `semaphore` this could look like this:

```c++
#include <atomic>
#include <functional>
#include <iostream>

template <typename A, typename Fn>
void cxx_atomic_wait(A *, Fn &&fn, std::memory_order mo);

struct semaphore {
    std::atomic<std::ptrdiff_t> a_{};

    bool try_acquire_impl(std::ptrdiff_t &value) {
        do {
            if (value == 0) {
                return false;
            }
        } while (!a_.compare_exchange_weak(value, value - 1,
                                           std::memory_order_acquire,
                                           std::memory_order_relaxed));
        return true;
    }

    static constexpr auto acquire_initial_load_mo = std::memory_order_relaxed;

    bool try_acquire() {
        auto old = a_.load(acquire_initial_load_mo);
        return try_acquire_impl(old);
    }

    void acquire() {
        using namespace std::placeholders;
        cxx_atomic_wait(&a_, std::bind(&semaphore::try_acquire_impl, this, _1),
                        acquire_initial_load_mo);
    }
};
```

Just for fun I could even use `std::bind` in `acquire()` to really make it clear that no custom (lambda) logic is needed there.

Also, the initial load can be `relaxed`, and `compare_exchange_weak` can be used instead of `compare_exchange_strong` since it's called in a loop anyway.

Reply:
I am still not convinced. In my opinion, it is simply a choice of on which abstraction we should build our `counting_semaphore` and `latch` on. I think your suggestion is that we should build these things on top of the internal `__cxx_atomic_wait`, `contention_state`, `__libcpp_thread_poll_with_backoff`. In my opinion, we should simply build it on top of `atomic`.

If you look at the current form of the PR, it only happens to work because:
- `counting_semaphore` uses `__atomic_base<ptrdiff_t>` and it contains `__cxx_atomic_impl<ptrdiff_t>`
- `__cxx_atomic_contention_t`  happens to be  `__cxx_atomic_impl<__cxx_contention_t>`
- `__cxx_contention_t` happens to be `int64_t`
- `ptrdiff_t` happens to be `int64_t`
- thus  `__atomic_base<ptrdiff_t>` happens to contain a `__cxx_atomic_contention_t`
- thus  `auto __monitor = std::__libcpp_atomic_monitor(__a_);` happens to call `__libcpp_atomic_monitor` the overload that takes `__cxx_atomic_contention_t` which uses its input (the `atomic` variable inside `semaphore` ) directly. If this type does not match, the fallback overload would use the global contention table and our write to the `monitor` variable would not update the `atomic` variable at all. (https://github.com/llvm/llvm-project/blob/main/libcxx/src/atomic.cpp#L163C46-L163C69)

I think, we should avoid messing around with these internal APIs, especially not to mess around the backing off functions that are dealing with contention state.  Instead, we could just use `atomic_base::{wait, notify_all, compare_exchange_weak, fetch_add, fetch_sub, __whatever_else_that_is_needed}`, so that our abstraction is built on top of something that is working and clear about its behaviour. If it is not working, then we know that we have a bug inside `atomic` and we would just fix the bug there. Instead of now, we are thinking does `A`, `B`, `C` that uses these internal functions have the same problem and we need to fix `A`, `B` and `C`.

then `counting_semaphore` and `latch` could be something like
```
struct semaphore{
   atomic_base<int> a_;

    void acquire() {
        auto old = a_.load(std::memory_order_relaxed);
        while(true) {
            if (old == 0) {
                a_.wait(old, std::memory_order_relaxed);
                old = a_.load(std::memory_order_relaxed);
            } else if (a_.compare_exchange_weak(old , old - 1, std::memory_order_acquire, std::memory_order_relaxed)){
                break;
            }
        }
    }

    void release(int update = 1) {
        auto old = a_.fetch_add(update, std::memory_order_release);
        if (old == 0) {
            a_.notify_all();
        }
    }
};


struct latch {
   atomic_base<int> a_;

    void count_down(int c = 1) {
        auto old = a_.fetch_sub(c, std::memory_order_release);
        if (old == 0) {
            a_.notify_all();
        }
    }

    void wait() {
        a_.wait(0, std::memory_order_acquire);
    }
};
```
And we don't care what pattern it is using, it is an implementation detail of `atomic_base::wait`. yes maybe the interface of `atomic` is insufficient to implement all the functions but we could add them as we need. I think working on higher level abstractions is making the code easier to maintain.

Reply:
> I am still not convinced. In my opinion, it is simply a choice of on which abstraction we should build our `counting_semaphore` and `latch` on. I think your suggestion is that we should build these things on top of the internal `__cxx_atomic_wait`, `contention_state`, `__libcpp_thread_poll_with_backoff`. In my opinion, we should simply build it on top of `atomic`.

My suggestion is to build on `atomic` and additionally on a fixed `__cxx_atomic_wait`. The higher level `semaphore` and `latch` shouldn't have to know about `contention_state` or `__libcpp_thread_poll_with_backoff`, I agree completely. You don't want to think about those internals when implementing high level logic.

`__cxx_atomic_wait` is the higher level abstraction you want, I believe. One can also think about it as the normal `.wait()`, but instead of a hardcoded condition (does `this->load(order)` compare equal to `old`?) you can put in any condition you want. So additionally to

```c++
void wait(T old, std::memory_order order = std::memory_order::seq_cst) const noexcept;
```

you have (`std::function` for clarity):

```c++
void __wait(std::function<bool(typename std::atomic<T>::value_type &old)> condition, std::memory_order order = std::memory_order::seq_cst) const noexcept; // this is basically (a fixed) __cxx_atomic_wait
```

This makes writing `semaphore` and `latch` easier, because you can just put in the `try_acquire()`s as the condition.

Reply:
> I played around a bit with this approach back in the day if you are interested: https://github.com/jiixyj/llvm-project/commit/bc11102a7707b8b8f19307984b732ae1f6933a6a

Just FYI, I'll try and rebase this on current main later today. That way we can better compare with this current PR.

Reply:
I think we are coming close to an agreement now.  yes we both agree to work on the higher level abstraction instead of trying fighting the backoff functions. yes having another function that takes a predicate can be useful too. maybe call it `wait_if_not`   (like `find` vs `find_if`). 

For `semaphore` case, I don't think we have to reuse `try_acquire` in `acquire` as I think they are a bit different.
1. We don't need a loop in `try_acquire`. can we just change it to oneoff thing?
2. we need `compare_exchange_strong` in `try_acquire` and `compare_exchange_weak` in `acquire`

something like
```
bool try_acquire() {
    auto __old = __a_.load(memory_order_acquire);
    return __old != 0 && __a_.compare_exchange_strong(__old, __old - 1, memory_order_acquire, memory_order_relaxed);
}

void acquire() {
    auto __old = __a_.load(std::memory_order_relaxed);
    while (true) {
      if (__old == 0) {
        __a_.wait(__old, std::memory_order_relaxed);
        __old = __a_.load(std::memory_order_relaxed);
      } else if (__a_.compare_exchange_weak(__old, __old - 1, std::memory_order_acquire, std::memory_order_relaxed)) {
        break;
      }
}
```

here is just a small refactoring of `semaphore` `latch` `barrier` to just use existing `atomic`'s API
https://github.com/huixie90/llvm-project/commit/8a4b98a2bdaab954c13a35bbd9b07e1c44ca69fe

Going back to the original bug. I don't think the refactoring of all this solved anything. The original lost wake up bug was not in `semaphore`, or `latch` etc...  It in the `atomic::wait`'s `backoff` function:
```
auto const __monitor = std::__libcpp_atomic_monitor(__a);
if (__test_fn())
      return true;
std::__libcpp_atomic_wait(__a, __monitor);
```
when this function is called, it is possible that the `atomic` variable has already been changed and `__monitor` is the new value. so we are waiting for a different value now. (and also there is no happens-before so __test_fn could be evaluated before taking the monitor). so basically we need to fix this function.
(also we need to careful about `__libcpp_atomic_monitor` returns different thing for `atomic<int64t>` and everything else

Reply:
> For `semaphore` case, I don't think we have to reuse `try_acquire` in `acquire` as I think they are a bit different.
> 
>     1. We don't need a loop in `try_acquire`. can we just change it to oneoff thing?
> 
>     2. we need `compare_exchange_strong` in `try_acquire` and `compare_exchange_weak` in `acquire`

Yes, you don't need a loop in `try_acquire()`. But then, you don't need `compare_exchange_strong` either, as `weak` will do. Imagine this sequence where a consumer calls this version of `try_acquire()` (which doesn't use a loop and uses `compare_exchange_strong`), and there is a concurrent producer:

```
producer         consumer         sem
-------------------------------------
                                   42
-------------------------------------
                  old=42           42
-------------------------------------
release(1)                         43
-------------------------------------
                  tries
                  strong           43
                  cas: fails!
-------------------------------------
                  returns false
```

...so you might then as well have used `compare_exchange_weak` as this is allowed to spuriously fail.

So you _can_ implement `acquire()` even in terms of this "weaker" `try_acquire()` -- you just have to make sure that any waiting happens based on the value "0" of the semaphore like this:

```c++
struct semaphore {
    std::atomic<std::ptrdiff_t> a_{};

    bool try_acquire_impl(std::ptrdiff_t &value) {
        // if `try_acquire_impl` returns `false`, `value` must be 0!
        return (value != 0 &&
                a_.compare_exchange_weak(value, value - 1,  //
                                         std::memory_order_acquire,
                                         std::memory_order_relaxed)) ||
               (value = 0, false);
    }

    static constexpr auto acquire_initial_load_mo = std::memory_order_relaxed;

    bool try_acquire() {
        auto old = a_.load(acquire_initial_load_mo);
        return try_acquire_impl(old);
    }

    void acquire() {
        using namespace std::placeholders;
        cxx_atomic_wait(&a_, std::bind(&semaphore::try_acquire_impl, this, _1),
                        acquire_initial_load_mo);
    }
};
```

Reply:
> when this function is called, it is possible that the atomic variable has already been changed and __monitor is the new value. so we are waiting for a different value now. (and also there is no happens-before so __test_fn could be evaluated before taking the monitor). so basically we need to fix this function.

Agreed! But I don't think the `__test_fn` can be reordered before taking the monitor, as taking the monitor is a "acquire" operation?

Reply:
> edit: Scratch the next part -- I'm not 100% sure, but it could well be that the condition for the `cxx_atomic_wait` has to be "strong". So if one has a "weak" `try_acquire()`, one has to implement the `acquire()` seperately, just as you wrote. I have to think about this some more :)

I think I got it! You basically can implement `try_acquire()` in three different ways, and the way you would implement `acquire()` in terms of that differs significantly:

1. "strong" `try_acquire()`: this only ever returns `false` if it has actually observed the semaphore value to be "0". You can implement `acquire()` in terms of that if you notify the eventcount if `old == 0` in `release()`.
2. "weak" `try_acquire()` with strong CAS (`return value != 0 && cas_strong(...);`): You could implement `acquire()` in terms of that, but then you would have to notify in any case, not only if `old == 0`! Example execution:

```
producer    consumer         sem        evc (eventcount)
-------------------------------------------
                              42          0
-------------------------------------------
             monitor=0        42          0
-------------------------------------------
             old=42           42          0
-------------------------------------------
release(1)                    43          0
-------------------------------------------
             strong
             CAS(42->41)      43          0
             fails
-------------------------------------------
evc.notify()                  43          1
^
this is
important!
-------------------------------------------
             evc.wait()       43          1
             ends
             as monitor
             doesn't
             match (0!=1)
-------------------------------------------
             loop restarts
             and strong CAS   42          1
             will succeed
             this time
```

3. "weak" `try_acquire()`, with weak CAS (`return value != 0 && cas_weak(...);`): You cannot implement `acquire()` in terms of that, because the weak CAS could just randomly decide to spuriously fail, and any notifications would get lost.

Sorry for the confused rambling, but I hope this kinda makes sense...